rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,149 @@
+PDF export. However, the way it works has downsides with performance,
+maintainability, and expanding it as a tool that can power higher-level
+features. The solution to those downsides is to have Kibana pages themselves
+become more capable at presenting themselves for screenshot capture reports.
+With the Screenshot Mode Service available, Reporting could drop the
+task that it currently has which hurt performance: wasted rendering that is
+replaced with custom styles that make the page "reportable."
+
+The technical advantage of having such as service also leads to making Kibana
+application pages "printable", in the sense that sending the page to a printer
+for a hard copy results in something more meaningful and specialized for paper
+than today's Kibana can guarantee. This isn't a big concern for Kibana since
+there isn't the expectation to improve printing Kibana, but that technical
+direction is appropriate for improving PDF report generation.
+
+# Detailed design
+
+The Screenshot Mode Service is a callable API available as dependency for
+Kibana applications.
+
+A method of the API tells the Application whether or not it should render
+itself to optimize for non-interactivity.
+
+In a future phase, the API might also tell the application more about the
+report, such as PDF page dimensions, or special layout types (print layout,
+etc).
+
+## Interface
+The `setupDeps.screenshotMode` object has a single purpose: tell the app if it
+should render in an optimized way for screenshot capture:
+
+```
+interface IScreenshotModeServiceSetup {
+  isScreenshotMode: () => boolean;
+}
+```
+
+Internally, this object is constructed from a class that refers to information
+sent via a custom proprietary header:
+
+```
+interface HeaderData {
+  'X-Screenshot-Mode': true
+}
+
+class ScreenshotModeServiceSetup implements IScreenshotModeServiceSetup {
+  constructor(rawData: HeaderData) {}
+  public isScreenshotMode (): boolean {}
+}
+```
+
+This works because the headless browser that opens the page can inject custom
+headers into the request. Teams can test how their app renders when loaded with
+this header using a new configuration setting, or a web debugging proxy, or
+some other tool that is TBD.
+
+# Alternatives
+
+- Print media query CSS
+  If applications UIs supported printability using `@media print`, and Kibana 
+  Reporting uses `page.print()` to capture the PDF, it would be easy for application 
+  developers to test, and prevent bugs showing up in the report.
+  
+  However, this solution doesn't include performance benefits of reducing objects 
+  in the headless browser memory: the headless browser still has to render the entire 
+  page as a "normal" render before it is able to call `page.print()`. No one sees the 
+  results of that initial render, so it is the same amount of wasted rendering cycles 
+  during report generation that we have today.
+
+# Adoption strategy
```

Comment:
What about the rollout strategy? I assume the old reporting plugin continues to be maintained alongside the new service, and it gets refactored to read from the screenshot service to that the existing app-specific logic can be be removed on a case-by-case basis as apps migrate?

Reply:
lets add something about this to this RFC, and yes in general that would be the approach I think. we would create a new report type (screenshot report type) and still maintain the old pdf/jpg report types until all applications using those migrate to the new report type

Reply:
Another thing I need to make clear: this doesn't replace anything in Reporting :) It's an add-on that gives wiring between Reporting and the Kibana UI.

I could see having a new report type which calls `page.print()` instead of `page.screenshot()` - that would be totally useful because text is selectable in the PDF if it was created with `page.print()`. There is a way to add that into the Reporting plugin (register a new Export Type) and trigger it in the report job with a new layout type. If we want to do that, it could live alongside the current export code.

Reply:
I updated the text to say that the Reporting team can work on the code that needs to be updated